lua/lspsaga/definition.lua

```diff
@@ -296,6 +296,155 @@ function def:goto_definition()
+        '[Lspsaga] response of request method textDocument/typeDefinition is nil',
+        vim.log.levels.WARN
+      )
+      return
+    end
+
+    local result
+    for _, res in pairs(results) do
+      if res and res.result then
+        result = res.result
+      end
+    end
+
+    if not result then
+      vim.notify(
+        '[Lspsaga] response of request method textDocument/typeDefinition is nil',
+        vim.log.levels.WARN
+      )
+      return
+    end
+
+    local res = get_uri_data(result)
+    if not res or not res.bufnr then
+      return
+    end
+
+    node.link = vim.uri_to_fname(res.uri)
+    local opts = {}
+    if stack_cap() == 0 then
+      local cur_winline = fn.winline()
+      local max_height = math.floor(vim.o.lines * 0.5)
+      local max_width = math.floor(vim.o.columns * 0.6)
+      opts = {
+        relative = 'cursor',
+        style = 'minimal',
+        no_override_size = true,
+        height = max_height,
+        width = max_width,
+      }
+      if vim.o.lines - opts.height - cur_winline < 0 then
+        vim.cmd('normal! zz')
+        local keycode = api.nvim_replace_termcodes('5<C-e>', true, false, true)
+        api.nvim_feedkeys(keycode, 'x', false)
+      end
+    else
+      opts = api.nvim_win_get_config(cur_winid)
+    end
+
+    local content_opts = {
+      contents = {},
+      enter = true,
+      bufnr = res.bufnr,
+      highlight = {
+        border = 'DefinitionBorder',
+        normal = 'DefinitionNormal',
+      },
+    }
+    --@deprecated when 0.9 release
+    if fn.has('nvim-0.9') == 1 and config.ui.title then
+      self:title_text(opts, node.link)
+    end
+
+    node.prev = api.nvim_get_current_win()
+    _, node.winid = window.create_win_with_border(content_opts, opts)
+    if config.symbol_in_winbar.enable then
+      api.nvim_win_set_var(node.winid, 'disable_winbar', true)
+    end
+    vim.opt_local.modifiable = true
+    node.bufnr = res.bufnr
+    node.wipe = res.wipe
+    --set the initail cursor pos
+    api.nvim_win_set_cursor(node.winid, { res.range.start.line + 1, res.range.start.character })
+    vim.cmd('normal! zt')
+    push(node)
+
+    apply_aciton_keys({ res.range.start.line, res.range.start.character })
+
+    api.nvim_create_autocmd('WinClosed', {
+      group = group,
+      buffer = node.bufnr,
+      callback = function(opt)
+        local curwinid = tonumber(opt.match)
+        local index = find_node(curwinid)
+        if index then
+          local wins = fn.win_findbuf(opt.buf)
+          wins = vim.tbl_filter(function(k)
+            return find_node(k)
+          end, wins)
+
+          if #wins == 1 and wins[1] == curwinid then
+            for _, map in pairs(config.definition) do
+              api.nvim_buf_del_keymap(opt.buf, 'n', map)
+            end
+          end
+
+          remove(index)
+        end
+      end,
+    })
+  end)
+end
+
+-- override the default the type defintion handler
+function def:goto_type_definition()
```

Comment:
these also duplicate codes. can we abstract a public function like `def:goto_definition(method)`

Reply:
Thanks for the review, indeed that will cut down the duplicates even more, i have used `strings.format` to pass the method in the type of handlers required for the method. Let me know if you have a better way to handle the strings, I preferred to pass only the second part of the handler endpoint for simplicity (`definition` instead of `textDocument/definition`)

Reply:
use number will be better  in command paramter it just take 1 byte. a local function which have method table. this table will marked by gc when the invoked function finished.   While numbers as parameters. make the meaning unclear. But it's more convenient.

```lua

--commandl.lua
peek_defintion(1)  1 or 2

-- definition.lua
local function get_method(index)
  local tbl = { method1, method2}
  return tbl[index]
end
```